File: shared/src/commonMain/kotlin/pe/ecolecta/domain/ProveedorQr.kt
package pe.ecolecta.domain

/**
 * El QR de un proveedor codifica únicamente su [Proveedor.codigo] con un prefijo de formato: nunca datos
 * sensibles (DNI, nombre, etc.). Se usa el código porque es lo único que identifica la MISMA ficha en
 * todos lados: el id local es un UUID distinto en cada celular y el panel web usa su propio id entero.
 * Mismo formato que el panel (web: App\Domain\Proveedores\ProveedorQr).
 *
 * Los QR anteriores (`ECOLECTA:PROVEEDOR:<id>`) se siguen leyendo para no invalidar los ya impresos.
 */
private const val PREFIJO_QR_PROVEEDOR = "ECOLECTA:PROVEEDOR:"
private const val PREFIJO_QR_CODIGO = "ECOLECTA:PROVEEDOR:CODIGO:"

fun generarQrProveedor(proveedorCodigo: String): String = "$PREFIJO_QR_CODIGO$proveedorCodigo"

/** Qué identifica un QR de proveedor leído. */
sealed interface ReferenciaQrProveedor {
    data class PorCodigo(val codigo: String) : ReferenciaQrProveedor

    /** Formato anterior: id del celular que lo generó (o del panel web, si era un número). */
    data class PorIdAntiguo(val id: String) : ReferenciaQrProveedor
}

fun leerQrProveedor(contenido: String): ReferenciaQrProveedor? {
    val valor = contenido.trim()
    return when {
        valor.startsWith(PREFIJO_QR_CODIGO) ->
            valor.removePrefix(PREFIJO_QR_CODIGO).ifBlank { null }?.let { ReferenciaQrProveedor.PorCodigo(it) }
        valor.startsWith(PREFIJO_QR_PROVEEDOR) ->
            valor.removePrefix(PREFIJO_QR_PROVEEDOR).ifBlank { null }?.let { ReferenciaQrProveedor.PorIdAntiguo(it) }
        else -> null
    }
}

File: web/app/Domain/Proveedores/ProveedorQr.php
<?php

namespace App\Domain\Proveedores;

/**
 * El QR de un proveedor codifica únicamente su código (nunca datos sensibles como DNI o nombre), con un
 * prefijo de formato para distinguir "QR ajeno a Ecolecta" de "proveedor inexistente".
 *
 * Se usa el código porque es lo único que identifica la misma ficha en el panel y en cada celular: los
 * ids (entero aquí, UUID local en la app) son distintos en cada lado. Mismo formato que la app móvil
 * (ver domain/ProveedorQr.kt). Los QR antiguos con el id del panel se siguen leyendo.
 */
final class ProveedorQr
{
    private const string PREFIJO = 'ECOLECTA:PROVEEDOR:';

    private const string PREFIJO_CODIGO = 'ECOLECTA:PROVEEDOR:CODIGO:';

    public static function generar(string $codigo): string
    {
        return self::PREFIJO_CODIGO.$codigo;
    }

    public static function extraerCodigo(string $contenido): ?string
    {
        $valor = trim($contenido);

        if (! str_starts_with($valor, self::PREFIJO_CODIGO)) {
            return null;
        }

        $codigo = substr($valor, strlen(self::PREFIJO_CODIGO));

        return $codigo !== '' ? $codigo : null;
    }

    /** Formato anterior (id del panel): solo para leer QR ya impresos. */
    public static function extraerId(string $contenido): ?int
    {
        $valor = trim($contenido);

        if (! str_starts_with($valor, self::PREFIJO) || str_starts_with($valor, self::PREFIJO_CODIGO)) {
            return null;
        }

        $id = substr($valor, strlen(self::PREFIJO));

        return ctype_digit($id) ? (int) $id : null;
    }
}
